py-polars/polars/schema.py

```diff
@@ -0,0 +1,61 @@
+from __future__ import annotations
+
+from collections import OrderedDict
+from typing import TYPE_CHECKING, Iterable, Mapping
+
+if TYPE_CHECKING:
+    from polars.datatypes import DataType
+
+    BaseSchema = OrderedDict[str, DataType]
+else:
+    # Python 3.8 does not support generic OrderedDict at runtime
+    BaseSchema = OrderedDict
+
+
+class Schema(BaseSchema):
+    """
+    Ordered mapping of column names to their data type.
+
+    Parameters
+    ----------
+    schema
+        The schema definition given by column names and their associated *instantiated*
+        Polars data type. Accepts a mapping or an iterable of tuples.
+
+    Examples
+    --------
+    Define a schema by passing *instantiated* data types.
+
+    >>> schema = pl.Schema({"foo": pl.Int8(), "bar": pl.String()})
+    >>> schema
+    Schema({'foo': Int8, 'bar': String})
+
+    Access the data type associated with a specific column name.
+
+    >>> schema["foo"]
+    Int8
+
+    Access various schema properties using the `names`, `dtypes`, and `len` methods.
+
+    >>> schema.names()
+    ['foo', 'bar']
+    >>> schema.dtypes()
+    [Int8, String]
+    >>> schema.len()
+    2
+    """
+
+    def __init__(self, schema: Mapping[str, DataType] | Iterable[tuple[str, DataType]]):
+        super().__init__(schema)
+
+    def names(self) -> list[str]:
+        """Get the column names of the schema."""
+        return list(self.keys())
+
+    def dtypes(self) -> list[DataType]:
+        """Get the data types of the schema."""
+        return list(self.values())
+
+    def len(self) -> int:
```

Comment:
We don't need this one; all Python maps/lists/etc support len(object) as being idiomatic - none expose an explicit method 🤔

Reply:
Thanks for the input. I'm going to keep the `len()` method - it's not strictly necessary as you say, but I think it fits well in the general chaining API that Polars offers. `len(schema)` still works so it doesn't really hurt I think.

Reply:
Some bonus colour: Guido explaining why `len(x)` is standard and `x.len()` isn't ;)
https://mail.python.org/pipermail/python-3000/2006-November/004643.html
```
(b) When I read code that says len(x) I *know* that it is asking for
the length of something. This tells me two things: the result is an
integer, and the argument is some kind of container. To the contrary,
when I read x.len(), I have to already know that x is some kind of
container implementing an interface or inheriting from a class that
has a standard len(). Witness the confusion we occasionally have when
a class that is not implementing a mapping has a get() or keys()
method, or something that isn't a file has a write() method.

Saying the same thing in another way, I see 'len' as a built-in
*operation*. I'd hate to lose that. I can't say for sure whether you
meant that or not, but 'def len(self): ...' certainly sounds like you
want to demote it to an ordinary method. I'm strongly -1 on that.
```

Reply:
Always nice to read some vintage Guido 😄